server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java

```diff
@@ -197,20 +200,41 @@ private PublishWithJoinResponse handleIncomingPublishRequest(BytesTransportReque
                         incomingState.stateUUID(),
                         request.bytes().length()
                     );
-                    final PublishWithJoinResponse response = acceptState(incomingState);
-                    lastSeenClusterState.compareAndSet(lastSeen, incomingState);
-                    return response;
+                    acceptState(
+                        incomingState,
+                        transportChannel,
+                        (acceptedState) -> lastSeenClusterState.compareAndSet(lastSeen, acceptedState)
```

Comment:
`onSuccess` will end up being called by the `CLUSTER_COORDINATION` thread only after a successful apply, so I wonder if we need to use compare-and-set here, I don't _think_ it's possible for `lastSeenCluster` to have changed between this being dispatched AND the new state being successfully applied. I would think either
- the state _has not_ changed and `onSuccess` is called correctly
- the state _has_ changed and the apply fails due to the version check, so `onSuccess` is not called

Reply:
The reason I raise this is because we could do away with the `onSuccess` callback if we were able to safely call `lastSeenClusterState.set` for both the full and diff payloads.

Reply:
I think it can change in between, dispatch and execution, yes - there could be another update in flight when we read the value when then completes and updates it before we get to run.

I am however not sure whether this matters. I'll think about this a little more.

Reply:
This one still concerns me a bit... the task of applying an update is done in three steps
1. deserialise payload
2. apply payload if valid
3. send response

only (2) happens in the mutex. So you could have the following interleaving for payload (a - version 7, b - version 8)

1. a-1
2. b-1
3. a-2 (succeeds, bump version to 7)
4. b-2 (successds, bump version to 8)
5. b-3 (set lastSeen to 8)
6. c-3 (set lastSeen to 7) - this means the next diff would be applied to 7, and a 7/9 hybrid would be applied locally?

The above could only happen if there were multiple threads in the `CLUSTER_COORDINATION` pool, but you said that is configurable.

We could put an additional check to only bump lastSeen when term and version are >= existing term and version?

Do we run tests with `CLUSTER_COORDINATION` size > 1 ?

Reply:
The sequence of 6 steps is indeed something that could have happened prior to #83576 and can again happen after this change - we could end up having applied state version 8 but with `lastSeen` at state version 7. But that's ok, each diff includes a UUID which identifies the base version which we check here:

https://github.com/elastic/elasticsearch/blob/9ba5651e7478216f17d81b63385ddf6db2331f7f/server/src/main/java/org/elasticsearch/cluster/ClusterState.java#L1170-L1172

That `IncompatibleClusterStateVersionException` is caught and handled here:

https://github.com/elastic/elasticsearch/blob/638a45009cb76059dd914c70eb4aa7e777768da2/server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java#L473-L483

IOW if we receive a diff between versions 8 & 9 but `lastSeen` is at version 7 then we reject the diff and the master sends us the full state at version 9 for us to apply. Somewhat inefficient for sure but still correct.

This case will be exercised in `org.elasticsearch.cluster.coordination.CoordinatorTests` in a `cluster.runRandomly()` call, but with rather low probability I think. We could try adding a test there which specifically checks this case.

Reply:
Great! not a problem we need to solve here then :)